# Segmenting and Clustering Neighborhoods in Toronto - Project



# Part 1

Importing Needed packages

In [1]:
import pandas as pd
import numpy as np
import requests

Get the HTML page of Wiki, and using read_html we convert the html data into list of Data frame objects.

In [2]:
wiki = 'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'
wiki_page = requests.get(wiki)
wiki_page

<Response [200]>

In [3]:
wiki_raw = pd.read_html(wiki_page.content, header = 0)[0]
wiki_raw

Postal Code           Borough  \
0           M1A      Not assigned   
1           M2A      Not assigned   
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
..          ...               ...   
175         M5Z      Not assigned   
176         M6Z      Not assigned   
177         M7Z      Not assigned   
178         M8Z         Etobicoke   
179         M9Z      Not assigned   

                                         Neighbourhood  
0                                         Not assigned  
1                                         Not assigned  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
..                                                 ...  
175                                       Not assigned  
176                                       Not assigned  
177                                       Not assigned  
178  Mimico NW, The Queensway West, South of Bloor,...  
179                                       Not assigned  

[180 rows x 3 columns]

Dropping Neighbourhood which are not assigned

In [4]:
df = wiki_raw[wiki_raw["Neighbourhood"] != "Not assigned"]
df

Postal Code           Borough  \
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
5           M6A        North York   
6           M7A  Downtown Toronto   
..          ...               ...   
160         M8X         Etobicoke   
165         M4Y  Downtown Toronto   
168         M7Y      East Toronto   
169         M8Y         Etobicoke   
178         M8Z         Etobicoke   

                                         Neighbourhood  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
5                     Lawrence Manor, Lawrence Heights  
6          Queen's Park, Ontario Provincial Government  
..                                                 ...  
160      The Kingsway, Montgomery Road, Old Mill North  
165                               Church and Wellesley  
168  Business reply mail Processing Centre, South C...  
169  Old Mill South, King's Mill Park, Sunnylea, Hu...  
178  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

Grouping the records based on Postal Code

In [5]:
df = df.groupby(['Postal Code']).head()
df

Postal Code           Borough  \
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
5           M6A        North York   
6           M7A  Downtown Toronto   
..          ...               ...   
160         M8X         Etobicoke   
165         M4Y  Downtown Toronto   
168         M7Y      East Toronto   
169         M8Y         Etobicoke   
178         M8Z         Etobicoke   

                                         Neighbourhood  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
5                     Lawrence Manor, Lawrence Heights  
6          Queen's Park, Ontario Provincial Government  
..                                                 ...  
160      The Kingsway, Montgomery Road, Old Mill North  
165                               Church and Wellesley  
168  Business reply mail Processing Centre, South C...  
169  Old Mill South, King's Mill Park, Sunnylea, Hu...  
178  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

Checking for number of records where Neighbourhood is "Not assigned"

In [6]:
df.Neighbourhood.str.count("Not assigned").sum()

0

Reindex data

In [7]:
df = df.reset_index()
df

index Postal Code           Borough  \
0        2         M3A        North York   
1        3         M4A        North York   
2        4         M5A  Downtown Toronto   
3        5         M6A        North York   
4        6         M7A  Downtown Toronto   
..     ...         ...               ...   
98     160         M8X         Etobicoke   
99     165         M4Y  Downtown Toronto   
100    168         M7Y      East Toronto   
101    169         M8Y         Etobicoke   
102    178         M8Z         Etobicoke   

                                         Neighbourhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                            Regent Park, Harbourfront  
3                     Lawrence Manor, Lawrence Heights  
4          Queen's Park, Ontario Provincial Government  
..                                                 ...  
98       The Kingsway, Montgomery Road, Old Mill North  
99                                Church and Wellesley  
100  Business reply mail Processing Centre, South C...  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  
102  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 4 columns]

In [8]:
df.drop(['index'], axis = 'columns', inplace = True)
df

Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
..          ...               ...   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                         Neighbourhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                            Regent Park, Harbourfront  
3                     Lawrence Manor, Lawrence Heights  
4          Queen's Park, Ontario Provincial Government  
..                                                 ...  
98       The Kingsway, Montgomery Road, Old Mill North  
99                                Church and Wellesley  
100  Business reply mail Processing Centre, South C...  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  
102  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

In [9]:
df.shape

(103, 3)

Now we have 103 rows and 3 columns

# Part 2

Installing geocoder

In [10]:
pip install geocoder

Note: you may need to restart the kernel to use updated packages.


import geocoder

In [11]:
import geocoder

Tried the below approach for many times but it didn't work!, Changing the code cell to Text.

In [12]:
# initialize your variable to None
lat_lng_coords = None

postal_code = 'M3A'

# loop until you get the coordinates
while(lat_lng_coords is None):
  g = geocoder.google('{}, Toronto, Ontario'.format(postal_code))
  lat_lng_coords = g.latlng

latitude = lat_lng_coords[0]
longitude = lat_lng_coords[1]

KeyboardInterrupt: 

In [13]:
url = 'http://cocl.us/Geospatial_data'

Importing the CSV file from the URL

In [14]:
df_geo = pd.read_csv(url)
df_geo.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

Checking dtypes of our columns

In [15]:
df_geo.dtypes

Postal Code     object
Latitude       float64
Longitude      float64
dtype: object

In [16]:
df.dtypes

Postal Code      object
Borough          object
Neighbourhood    object
dtype: object

Shapes of both our Dataframes

In [17]:
df.shape

(103, 3)

In [18]:
df_geo.shape

(103, 3)

Joining both df's and doing some cleaning

In [19]:
df = df.join(df_geo.set_index('Postal Code'), on='Postal Code')
df

Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
..          ...               ...   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                         Neighbourhood   Latitude  Longitude  
0                                            Parkwoods  43.753259 -79.329656  
1                                     Victoria Village  43.725882 -79.315572  
2                            Regent Park, Harbourfront  43.654260 -79.360636  
3                     Lawrence Manor, Lawrence Heights  43.718518 -79.464763  
4          Queen's Park, Ontario Provincial Government  43.662301 -79.389494  
..                                                 ...        ...        ...  
98       The Kingsway, Montgomery Road, Old Mill North  43.653654 -79.506944  
99                                Church and Wellesley  43.665860 -79.383160  
100  Business reply mail Processing Centre, South C...  43.662744 -79.321558  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  43.636258 -79.498509  
102  Mimico NW, The Queensway West, South of Bloor,...  43.628841 -79.520999  

[103 rows x 5 columns]

In [20]:
df = df.reset_index()

In [21]:
df.drop(['index'], axis = 'columns', inplace = True)

In [22]:
df.head()

Postal Code           Borough                                Neighbourhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  
0  43.753259 -79.329656  
1  43.725882 -79.315572  
2  43.654260 -79.360636  
3  43.718518 -79.464763  
4  43.662301 -79.389494

In [23]:
df.index.name = 'index'

In [24]:
df.head()

Postal Code           Borough  \
index                                 
0             M3A        North York   
1             M4A        North York   
2             M5A  Downtown Toronto   
3             M6A        North York   
4             M7A  Downtown Toronto   

                                     Neighbourhood   Latitude  Longitude  
index                                                                     
0                                        Parkwoods  43.753259 -79.329656  
1                                 Victoria Village  43.725882 -79.315572  
2                        Regent Park, Harbourfront  43.654260 -79.360636  
3                 Lawrence Manor, Lawrence Heights  43.718518 -79.464763  
4      Queen's Park, Ontario Provincial Government  43.662301 -79.389494

In [25]:
df.shape

(103, 5)

We get 103 rows as expected when we do a inner join, so we have good data.

# Part 3

Drawing inspiration from the previous lab where we cluster the neighbourhood of NYC, We cluster Toronto based on the similarities of the venues categories using Kmeans clustering and Foursquare API.

In [46]:
import geocoder
import geopy
from geopy.geocoders import Nominatim

In [47]:
address = 'Toronto, Ontario'

geolocator = Nominatim(user_agent="toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The coordinates of Toronto are {}, {}.'.format(latitude, longitude))

The coordinates of Toronto are 43.6534817, -79.3839347.


Let's visualize the map of Toronto

In [39]:
pip install folium

Note: you may need to restart the kernel to use updated packages.


In [48]:
import folium

In [49]:
# Creating the map of Toronto
map_Toronto = folium.Map(location=[latitude, longitude], zoom_start=11)

# adding markers to map
for latitude, longitude, borough, neighbourhood in zip(df['Latitude'], df['Longitude'], df['Borough'], df['Neighbourhood']):
    label = '{}, {}'.format(neighbourhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [latitude, longitude],
        radius=5,
        popup=label,
        color='red',
        fill=True
        ).add_to(map_Toronto)  
    
map_Toronto

Define Foursquare parametrs

In [54]:
CLIENT_ID = 'IEO3BCOMPMVDENSQXFGQOO21FUGTKIV1ZX01F4LNKLV0PK25' # your Foursquare ID
CLIENT_SECRET = 'XNJPG4KKQYJR2ZJM1PFS1FBAOM3APZ32TT2KDDQ3QNBPWCV0' # your Foursquare Secret
VERSION = '20180604' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: IEO3BCOMPMVDENSQXFGQOO21FUGTKIV1ZX01F4LNKLV0PK25
CLIENT_SECRET:XNJPG4KKQYJR2ZJM1PFS1FBAOM3APZ32TT2KDDQ3QNBPWCV0


Explore the data, and get the venues in 500 meters range from our first entry and get the url

In [82]:

neighborhood_latitude = df.loc[0, 'Latitude'] # neighborhood latitude value
neighborhood_longitude = df.loc[0, 'Longitude'] # neighborhood longitude value

neighborhood_name = df.loc[0, 'Neighbourhood'] # neighborhood name

print('Latitude and longitude values of {} are {}, {}.'.format(neighborhood_name, 
                                                               neighborhood_latitude, 
                                                               neighborhood_longitude))

Latitude and longitude values of Parkwoods are 43.7532586, -79.3296565.


In [83]:
LIMIT = 100
radius = 500
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    neighborhood_latitude, 
    neighborhood_longitude, 
    radius, 
    LIMIT)
url

'https://api.foursquare.com/v2/venues/explore?&client_id=IEO3BCOMPMVDENSQXFGQOO21FUGTKIV1ZX01F4LNKLV0PK25&client_secret=XNJPG4KKQYJR2ZJM1PFS1FBAOM3APZ32TT2KDDQ3QNBPWCV0&v=20180604&ll=43.7532586,-79.3296565&radius=500&limit=100'

In [58]:
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5fb869d3b8c07577f7efd64b'},
 'response': {'warning': {'text': "There aren't a lot of results near you. Try something more general, reset your filters, or expand the search area."},
  'headerLocation': 'Parkwoods - Donalda',
  'headerFullLocation': 'Parkwoods - Donalda, Toronto',
  'headerLocationGranularity': 'neighborhood',
  'totalResults': 2,
  'suggestedBounds': {'ne': {'lat': 43.757758604500005,
    'lng': -79.32343823984928},
   'sw': {'lat': 43.7487585955, 'lng': -79.33587476015072}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '4e8d9dcdd5fbbbb6b3003c7b',
       'name': 'Brookbanks Park',
       'location': {'address': 'Toronto',
        'lat': 43.751976046055574,
        'lng': -79.33214044722958,
        'labeledLatLngs': 

In [84]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

In [85]:
import json
from pandas.io.json import json_normalize

venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()

/opt/conda/envs/Python-3.7-main/lib/python3.7/site-packages/ipykernel/__main__.py:6: FutureWarning: pandas.io.json.json_normalize is deprecated, use pandas.json_normalize instead


name         categories        lat        lng
0  Brookbanks Park               Park  43.751976 -79.332140
1    Variety Store  Food & Drink Shop  43.751974 -79.333114

Generalize to obtain the venues from all neighbourhoods in Toronto

In [86]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighbourhood', 
                  'Neighbourhood Latitude', 
                  'Neighbourhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [87]:
toronto_venues = getNearbyVenues(names=df['Neighbourhood'],
                                   latitudes=df['Latitude'],
                                   longitudes=df['Longitude']
                                  )

Parkwoods
Victoria Village
Regent Park, Harbourfront
Lawrence Manor, Lawrence Heights
Queen's Park, Ontario Provincial Government
Islington Avenue, Humber Valley Village
Malvern, Rouge
Don Mills
Parkview Hill, Woodbine Gardens
Garden District, Ryerson
Glencairn
West Deane Park, Princess Gardens, Martin Grove, Islington, Cloverdale
Rouge Hill, Port Union, Highland Creek
Don Mills
Woodbine Heights
St. James Town
Humewood-Cedarvale
Eringate, Bloordale Gardens, Old Burnhamthorpe, Markland Wood
Guildwood, Morningside, West Hill
The Beaches
Berczy Park
Caledonia-Fairbanks
Woburn
Leaside
Central Bay Street
Christie
Cedarbrae
Hillcrest Village
Bathurst Manor, Wilson Heights, Downsview North
Thorncliffe Park
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Scarborough Village
Fairview, Henry Farm, Oriole
Northwood Park, York University
East Toronto, Broadview North (Old East York)
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
Kennedy Park, Ionview, East Birchmo

Checking the size of df

In [88]:
print(toronto_venues.shape)
toronto_venues.head()

(2167, 7)


Neighbourhood  Neighbourhood Latitude  Neighbourhood Longitude  \
0         Parkwoods               43.753259               -79.329656   
1         Parkwoods               43.753259               -79.329656   
2  Victoria Village               43.725882               -79.315572   
3  Victoria Village               43.725882               -79.315572   
4  Victoria Village               43.725882               -79.315572   

                    Venue  Venue Latitude  Venue Longitude  \
0         Brookbanks Park       43.751976       -79.332140   
1           Variety Store       43.751974       -79.333114   
2  Victoria Village Arena       43.723481       -79.315635   
3             Tim Hortons       43.725517       -79.313103   
4               Portugril       43.725819       -79.312785   

          Venue Category  
0                   Park  
1      Food & Drink Shop  
2           Hockey Arena  
3            Coffee Shop  
4  Portuguese Restaurant

Checking how many venues there are for each venue

In [89]:
toronto_venues.groupby('Neighbourhood').count()

Neighbourhood Latitude  \
Neighbourhood                                                             
Agincourt                                                             5   
Alderwood, Long Branch                                                7   
Bathurst Manor, Wilson Heights, Downsview North                      21   
Bayview Village                                                       4   
Bedford Park, Lawrence Manor East                                    22   
...                                                                 ...   
Willowdale, Willowdale West                                           5   
Woburn                                                                4   
Woodbine Heights                                                      9   
York Mills West                                                       2   
York Mills, Silver Hills                                              1   

                                                 Neighbourhood Longitude  \
Neighbourhood                                                              
Agincourt                                                              5   
Alderwood, Long Branch                                                 7   
Bathurst Manor, Wilson Heights, Downsview North                       21   
Bayview Village                                                        4   
Bedford Park, Lawrence Manor East                                     22   
...                                                                  ...   
Willowdale, Willowdale West                                            5   
Woburn                                                                 4   
Woodbine Heights                                                       9   
York Mills West                                                        2   
York Mills, Silver Hills                                               1   

                                                 Venue  Venue Latitude  \
Neighbourhood                                                            
Agincourt                                            5               5   
Alderwood, Long Branch                               7               7   
Bathurst Manor, Wilson Heights, Downsview North     21              21   
Bayview Village                                      4               4   
Bedford Park, Lawrence Manor East                   22              22   
...                                                ...             ...   
Willowdale, Willowdale West                          5               5   
Woburn                                               4               4   
Woodbine Heights                                     9               9   
York Mills West                                      2               2   
York Mills, Silver Hills                             1               1   

                                                 Venue Longitude  \
Neighbourhood                                                      
Agincourt                                                      5   
Alderwood, Long Branch                                         7   
Bathurst Manor, Wilson Heights, Downsview North               21   
Bayview Village                                                4   
Bedford Park, Lawrence Manor East                             22   
...                                                          ...   
Willowdale, Willowdale West                                    5   
Woburn                                                         4   
Woodbine Heights                                               9   
York Mills West                                                2   
York Mills, Silver Hills                                       1   

                                                 Venue Category  
Neighbourhood                                                    
Agincourt                                                     5  
Alderwood, Long Branch                

In [90]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighbourhood'] = toronto_venues['Neighbourhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

Neighbourhood  Accessories Store  Afghan Restaurant  Airport  \
0         Parkwoods                  0                  0        0   
1         Parkwoods                  0                  0        0   
2  Victoria Village                  0                  0        0   
3  Victoria Village                  0                  0        0   
4  Victoria Village                  0                  0        0   

   Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0                   0             0               0                0   
1                   0             0               0                0   
2                   0             0               0                0   
3                   0             0               0                0   
4                   0             0               0                0   

   Airport Terminal  American Restaurant  ...  Train Station  \
0                 0                    0  ...              0   
1                 0                    0  ...              0   
2                 0                    0  ...              0   
3                 0                    0  ...              0   
4                 0                    0  ...              0   

   Turkish Restaurant  Vegetarian / Vegan Restaurant  Video Game Store  \
0                   0                              0                 0   
1                   0                              0                 0   
2                   0                              0                 0   
3                   0                              0                 0   
4                   0                              0                 0   

   Vietnamese Restaurant  Warehouse Store  Wine Bar  Wings Joint  \
0                      0                0         0            0   
1                      0                0         0            0   
2                      0                0         0            0   
3                      0                0         0            0   
4                      0                0         0            0   

   Women's Store  Yoga Studio  
0              0            0  
1              0            0  
2              0            0  
3              0            0  
4              0            0  

[5 rows x 278 columns]

In [67]:
toronto_onehot.shape

(2167, 278)

In [91]:
toronto_grouped = toronto_onehot.groupby('Neighbourhood').mean().reset_index()
toronto_grouped.head()

Neighbourhood  Accessories Store  \
0                                        Agincourt                0.0   
1                           Alderwood, Long Branch                0.0   
2  Bathurst Manor, Wilson Heights, Downsview North                0.0   
3                                  Bayview Village                0.0   
4                Bedford Park, Lawrence Manor East                0.0   

   Afghan Restaurant  Airport  Airport Food Court  Airport Gate  \
0                0.0      0.0                 0.0           0.0   
1                0.0      0.0                 0.0           0.0   
2                0.0      0.0                 0.0           0.0   
3                0.0      0.0                 0.0           0.0   
4                0.0      0.0                 0.0           0.0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0             0.0              0.0               0.0             0.000000   
1             0.0              0.0               0.0             0.000000   
2             0.0              0.0               0.0             0.000000   
3             0.0              0.0               0.0             0.000000   
4             0.0              0.0               0.0             0.045455   

   ...  Train Station  Turkish Restaurant  Vegetarian / Vegan Restaurant  \
0  ...            0.0                 0.0                            0.0   
1  ...            0.0                 0.0                            0.0   
2  ...            0.0                 0.0                            0.0   
3  ...            0.0                 0.0                            0.0   
4  ...            0.0                 0.0                            0.0   

   Video Game Store  Vietnamese Restaurant  Warehouse Store  Wine Bar  \
0               0.0                    0.0              0.0       0.0   
1               0.0                    0.0              0.0       0.0   
2               0.0                    0.0              0.0       0.0   
3               0.0                    0.0              0.0       0.0   
4               0.0                    0.0              0.0       0.0   

   Wings Joint  Women's Store  Yoga Studio  
0          0.0            0.0          0.0  
1          0.0            0.0          0.0  
2          0.0            0.0          0.0  
3          0.0            0.0          0.0  
4          0.0            0.0          0.0  

[5 rows x 278 columns]

In [92]:
toronto_grouped.shape

(96, 278)

Getting the top 10 for each neighbourhood

In [93]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [94]:
import numpy as np
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighbourhood'] = toronto_grouped['Neighbourhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighbourhood 1st Most Common Venue  \
0                                        Agincourt                Lounge   
1                           Alderwood, Long Branch           Pizza Place   
2  Bathurst Manor, Wilson Heights, Downsview North           Coffee Shop   
3                                  Bayview Village   Japanese Restaurant   
4                Bedford Park, Lawrence Manor East        Sandwich Place   

  2nd Most Common Venue      3rd Most Common Venue 4th Most Common Venue  \
0        Breakfast Spot  Latin American Restaurant          Skating Rink   
1        Sandwich Place                Coffee Shop                   Pub   
2                  Bank                   Pharmacy        Ice Cream Shop   
3                  Café         Chinese Restaurant                  Bank   
4    Italian Restaurant                Coffee Shop      Greek Restaurant   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0        Clothing Store           Event Space    Falafel Restaurant   
1              Pharmacy                   Gym      Greek Restaurant   
2         Shopping Mall           Bridal Shop        Sandwich Place   
3   Distribution Center              Dive Bar               Dog Run   
4       Thai Restaurant             Locksmith          Liquor Store   

     8th Most Common Venue      9th Most Common Venue 10th Most Common Venue  
0     Ethiopian Restaurant                Escape Room    Distribution Center  
1           Discount Store           Department Store           Dessert Shop  
2                    Diner  Middle Eastern Restaurant             Restaurant  
3         Doner Restaurant                 Donut Shop            Yoga Studio  
4  Comfort Food Restaurant                  Juice Bar                Butcher

Clustering Neighborhoods

In [95]:
# import k-means from clustering stage
from sklearn.cluster import KMeans

# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighbourhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10]

array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1], dtype=int32)

Merge the dataframe with the top 10 and the cluster for each neighbourhood

In [96]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = df

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighbourhood'), on='Neighbourhood')

toronto_merged.head() # check the last columns!

Postal Code           Borough  \
index                                 
0             M3A        North York   
1             M4A        North York   
2             M5A  Downtown Toronto   
3             M6A        North York   
4             M7A  Downtown Toronto   

                                     Neighbourhood   Latitude  Longitude  \
index                                                                      
0                                        Parkwoods  43.753259 -79.329656   
1                                 Victoria Village  43.725882 -79.315572   
2                        Regent Park, Harbourfront  43.654260 -79.360636   
3                 Lawrence Manor, Lawrence Heights  43.718518 -79.464763   
4      Queen's Park, Ontario Provincial Government  43.662301 -79.389494   

       Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
index                                                               
0                 0.0                  Park     Food & Drink Shop   
1                 1.0           Pizza Place          Hockey Arena   
2                 1.0           Coffee Shop                  Park   
3                 1.0        Clothing Store     Accessories Store   
4                 1.0           Coffee Shop           Yoga Studio   

      3rd Most Common Venue 4th Most Common Venue   5th Most Common Venue  \
index                                                                       
0               Yoga Studio   Dumpling Restaurant     Distribution Center   
1         French Restaurant           Coffee Shop   Portuguese Restaurant   
2                    Bakery                   Pub          Breakfast Spot   
3                  Boutique             Gift Shop  Furniture / Home Store   
4                      Bank              Beer Bar           Smoothie Shop   

      6th Most Common Venue       7th Most Common Venue 8th Most Common Venue  \
index                                                                           
0                  Dive Bar                     Dog Run      Doner Restaurant   
1              Intersection  Construction & Landscaping                 Diner   
2                      Café                     Theater                   Spa   
3               Event Space                 Coffee Shop         Women's Store   
4            Sandwich Place                        Café            Restaurant   

       9th Most Common Venue 10th Most Common Venue  
index                                                
0                 Donut Shop              Drugstore  
1       Ethiopian Restaurant            Escape Room  
2                    Brewery             Shoe Store  
3      Vietnamese Restaurant        Airport Service  
4         Chinese Restaurant  Portuguese Restaurant

In [97]:
toronto_merged[toronto_merged['Cluster Labels'].isnull()]

Postal Code      Borough                            Neighbourhood  \
index                                                                     
5             M9A    Etobicoke  Islington Avenue, Humber Valley Village   
52            M2M   North York                  Willowdale, Newtonbrook   
95            M1X  Scarborough                              Upper Rouge   

        Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
index                                                               
5      43.667856 -79.532242             NaN                   NaN   
52     43.789053 -79.408493             NaN                   NaN   
95     43.836125 -79.205636             NaN                   NaN   

      2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
index                                                                     
5                       NaN                   NaN                   NaN   
52                      NaN                   NaN                   NaN   
95                      NaN                   NaN                   NaN   

      5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
index                                                                     
5                       NaN                   NaN                   NaN   
52                      NaN                   NaN                   NaN   
95                      NaN                   NaN                   NaN   

      8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
index                                                                     
5                       NaN                   NaN                    NaN  
52                      NaN                   NaN                    NaN  
95                      NaN                   NaN                    NaN

Plot the clusters in the map

In [98]:
import matplotlib.cm as cm
import matplotlib.colors as colors

# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

toronto_merged_nonan = toronto_merged.dropna(subset=['Cluster Labels'])

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged_nonan['Latitude'], toronto_merged_nonan['Longitude'], toronto_merged_nonan['Neighbourhood'], toronto_merged_nonan['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[int(cluster-1)],
        fill=True,
        fill_color=rainbow[int(cluster-1)],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

Examine clusters

Cluster 1

In [99]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 0, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
index                                                           
0            North York             0.0                  Park   
21                 York             0.0                  Park   
35            East York             0.0          Intersection   
49           North York             0.0                Bakery   
61      Central Toronto             0.0                  Park   
64                 York             0.0                  Park   
66           North York             0.0                  Park   
85          Scarborough             0.0          Intersection   
91     Downtown Toronto             0.0                  Park   

      2nd Most Common Venue       3rd Most Common Venue 4th Most Common Venue  \
index                                                                           
0         Food & Drink Shop                 Yoga Studio   Dumpling Restaurant   
21            Women's Store                        Pool           Yoga Studio   
35                     Park           Convenience Store   Dumpling Restaurant   
49                     Park  Construction & Landscaping           Yoga Studio   
61              Swim School                    Bus Line           Yoga Studio   
64              Yoga Studio         Dumpling Restaurant   Distribution Center   
66        Convenience Store                 Yoga Studio   Dumpling Restaurant   
85               Playground                        Park                Bakery   
91                    Trail                  Playground          Tennis Court   

      5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
index                                                                     
0       Distribution Center              Dive Bar               Dog Run   
21                Drugstore        Discount Store   Distribution Center   
35      Distribution Center              Dive Bar               Dog Run   
49      Dumpling Restaurant              Dive Bar               Dog Run   
61               Donut Shop   Distribution Center              Dive Bar   
64                 Dive Bar               Dog Run      Doner Restaurant   
66      Distribution Center              Dive Bar               Dog Run   
85               Donut Shop        Discount Store   Distribution Center   
91               Donut Shop                 Diner        Discount Store   

      8th Most Common Venue 9th Most Common Venue       10th Most Common Venue  
index                                                                           
0          Doner Restaurant            Donut Shop                    Drugstore  
21                 Dive Bar               Dog Run             Doner Restaurant  
35         Doner Restaurant            Donut Shop                    Drugstore  
49         Doner Restaurant            Donut Shop                    Drugstore  
61                  Dog Run      Doner Restaurant                    Drugstore  
64               Donut Shop             Drugstore  Eastern European Restaurant  
66         Doner Restaurant            Donut Shop                    Drugstore  
85                 Dive Bar               Dog Run             Doner Restaurant  
91      Distribution Center              Dive Bar                      Dog Run

Cluster 2

In [100]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 1, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels       1st Most Common Venue  \
index                                                                 
1            North York             1.0                 Pizza Place   
2      Downtown Toronto             1.0                 Coffee Shop   
3            North York             1.0              Clothing Store   
4      Downtown Toronto             1.0                 Coffee Shop   
7            North York             1.0                         Gym   
...                 ...             ...                         ...   
98            Etobicoke             1.0                       River   
99     Downtown Toronto             1.0                 Coffee Shop   
100        East Toronto             1.0                 Pizza Place   
101           Etobicoke             1.0  Construction & Landscaping   
102           Etobicoke             1.0                         Gym   

      2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
index                                                                     
1              Hockey Arena     French Restaurant           Coffee Shop   
2                      Park                Bakery                   Pub   
3         Accessories Store              Boutique             Gift Shop   
4               Yoga Studio                  Bank              Beer Bar   
7                Beer Store   Japanese Restaurant           Coffee Shop   
...                     ...                   ...                   ...   
98                     Pool           Yoga Studio            Donut Shop   
99         Sushi Restaurant               Gay Bar   Japanese Restaurant   
100           Auto Workshop         Garden Center                Garden   
101          Baseball Field           Yoga Studio   Dumpling Restaurant   
102             Social Club       Supplement Shop                Bakery   

        5th Most Common Venue     6th Most Common Venue  \
index                                                     
1       Portuguese Restaurant              Intersection   
2              Breakfast Spot                      Café   
3      Furniture / Home Store               Event Space   
4               Smoothie Shop            Sandwich Place   
7                Tennis Court            Sandwich Place   
...                       ...                       ...   
98             Discount Store       Distribution Center   
99                 Restaurant  Mediterranean Restaurant   
100        Light Rail Station      Fast Food Restaurant   
101                  Dive Bar                   Dog Run   
102             Tanning Salon             Burrito Place   

            7th Most Common Venue 8th Most Common Venue  \
index                                                     
1      Construction & Landscaping                 Diner   
2                         Theater                   Spa   
3                     Coffee Shop         Women's Store   
4                            Café            Restaurant   
7                       Bike Shop   Sporting Goods Shop   
...                           ...                   ...   
98                       Dive Bar               Dog Run   
99                           Café           Men's Store   
100                Farmers Market            Comic Shop   
101              Doner Restaurant            Donut Shop   
102                  Burger Joint            Kids Store   

       9th Most Common Venue       10th Most Common Venue  
index                                                      
1       Ethiopian Restaurant                  Escape Room  
2                    Brewery                   Shoe Store  
3      Vietnamese Restaurant              Airport Service  
4         Chinese Restaurant        Portuguese Restaurant  
7                       Café                   Restaurant  
...                      ...                          ...  
98          Doner Restaurant                    Drugstore  
99               Yoga Studio            

Cluster 3

In [101]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 2, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
index                                                                           
45     North York             2.0   Martial Arts School           Yoga Studio   

      3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
index                                                                     
45      Dumpling Restaurant   Distribution Center              Dive Bar   

      6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
index                                                                     
45                  Dog Run      Doner Restaurant            Donut Shop   

      9th Most Common Venue       10th Most Common Venue  
index                                                     
45                Drugstore  Eastern European Restaurant

Cluster 4

In [102]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 3, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
index                                                                          
11     Etobicoke             3.0            Print Shop           Yoga Studio   

      3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
index                                                                     
11                Drugstore        Discount Store   Distribution Center   

      6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
index                                                                     
11                 Dive Bar               Dog Run      Doner Restaurant   

      9th Most Common Venue 10th Most Common Venue  
index                                               
11               Donut Shop    Dumpling Restaurant

Cluster 5

In [103]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 4, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
index                                                      
6      Scarborough             4.0  Fast Food Restaurant   

      2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
index                                                                     
6               Yoga Studio             Drugstore        Discount Store   

      5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
index                                                                     
6       Distribution Center              Dive Bar               Dog Run   

      8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
index                                                                     
6          Doner Restaurant            Donut Shop    Dumpling Restaurant